h/services/user_delete.py

```diff
@@ -1,70 +1,348 @@
-        annotation_delete_service: AnnotationDeleteService,
-    ):
-        self._db = db_session
-        self._annotation_delete_service = annotation_delete_service
 
-    def delete_user(self, user: User, requested_by: User, tag: str):
-        """
-        Delete a user with all their group memberships and annotations.
+    def __init__(self, db, job_queue, user_svc):
+        self.db = db
+        self.job_queue = job_queue
+        self.user_svc = user_svc
 
-        If a user owns groups with collaborators, meaning there are annotations
-        in the group that have been made by other users, the user is unassigned
-        as creator but the group persists.
-        """
-        self._db.execute(sa.delete(Token).where(Token.user == user))
+    def delete_user(self, user: User, requested_by: User, tag: str):
+        """Mark `user` as deleted and start purging their data in the background."""
 
-        # Delete all annotations
-        self._annotation_delete_service.delete_annotations(
-            annotations=self._db.query(Annotation).filter_by(userid=user.userid).all()
-        )
+        # Mark the user as deleted so they can't login anymore.
+        user.deleted = True
 
-        # Delete or remove our link to groups we've created
-        for group, annotations_by_other_users in self._db.execute(
-            # pylint:disable=not-callable
-            sa.select(Group, sa.func.count(Annotation.id))
-            .where(Group.creator == user)
-            .outerjoin(
-                Annotation,
-                sa.and_(
-                    Annotation.groupid == Group.pubid, Annotation.userid != user.userid
-                ),
+        # We can't just purge all the user's data right now because for users
+        # with a lot of data this takes too long for an HTTP request and the
+        # request times out.
+        #
+        # So add a job to purge the user's data in the background.
+        self.db.add(
+            Job(
+                name=Job.JobName.PURGE_USER,
+                priority=0,
+                tag="UserDeleteService.delete_user",
+                kwargs={"userid": user.userid},
             )
-            .group_by(Group.id)
-        ):
-            if annotations_by_other_users:
-                group.creator = None
-            else:
-                self._db.delete(group)
+        )
 
-        self._db.add(
+        # Record the deletion for record-keeping purposes.
+        self.db.add(
             UserDeletion(
                 userid=user.userid,
                 requested_by=requested_by.userid,
                 tag=tag,
                 registered_date=user.registered_date,
-                num_annotations=self._db.scalar(
-                    sa.select(
-                        sa.func.count(Annotation.id)  # pylint:disable=not-callable
+                num_annotations=self.db.scalar(
+                    select(
+                        func.count(Annotation.id)  # pylint:disable=not-callable
```

Comment:
Counting the `annotation` table here. Will this be fast enough? May want to change this to `AnnotationSlim` once back-filling it is finished.